nemo/core/classes/modelPT.py

```diff
@@ -676,6 +688,40 @@ def setup_optimizer_param_groups(self):
 
         self._optimizer_param_groups = param_groups
 
+    def setup(self, stage: Optional[str] = None):
+        """Called at the beginning of fit, validate, test, or predict.
+        This is called on every process when using DDP.
+
+        Args:
+            stage: fit, validate, test or predict
+        """
+        if stage == 'fit':
+            train_deferred_setup = (
+                'train_ds' in self._cfg
+                and self._cfg.train_ds is not None
+                and self._cfg.train_ds.get('defer_setup', False)
+            )
+            if self.train_dataloader() is None and train_deferred_setup:
```

Comment:
One thing im worried about is caching - it generally causes problems. But i suppose for the defer setup flag will mostly be fine since by default it is false and this codepath will not execute unless specified.

Reply:
One thing with the deffered flag is if you use it, you cannot register any artifact inside of setup_train/val/test methods. Which IMO is fine, we want to allow registration of artifacts only at init time.